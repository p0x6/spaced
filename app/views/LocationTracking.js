import React, { useState, useEffect } from 'react';
import {
  StyleSheet,
  View,
  Text,
  TouchableOpacity,
  Dimensions,
  BackHandler,
} from 'react-native';
import LocationServices from '../services/LocationService';
import BroadcastingServices from '../services/BroadcastingService';
import BackgroundGeolocation from '@mauron85/react-native-background-geolocation';

import { GetStoreData, SetStoreData } from '../helpers/General';
import OverlapScreen from './Overlap';
import SlidingUpPanel from 'rn-sliding-up-panel';
import ToggleSwitch from 'toggle-switch-react-native';
import { useNavigation } from '@react-navigation/native';

const width = Dimensions.get('window').width;

const LocationTracking = () => {
  const [isLogging, setIsLogging] = useState(false);
  const [panelHeight, setPanelHeight] = useState(180);
  const { navigate } = useNavigation();

  useEffect(() => {
    BackHandler.addEventListener('hardwareBackPress', handleBackPress);
    GetStoreData('PARTICIPATE')
      .then(isParticipating => {
        if (isParticipating === 'true') {
          setIsLogging(true);
          willParticipate();
        } else {
          setIsLogging(false);
        }
      })
      .catch(error => console.log(error));

    return BackHandler.removeEventListener(
      'hardwareBackPress',
      handleBackPress,
    );
  }, []);

  const handleBackPress = () => {
    BackHandler.exitApp(); // works best when the goBack is async
    return true;
  };

  const toExport = () => navigate('ExportScreen', {});

  const toImport = () => navigate('ImportScreen', {});

  const blacklistPlaces = () => navigate('BlacklistPlaces', {});

  const toNews = () => navigate('NewsScreen', {});

  const licenses = () => navigate('LicensesScreen', {});

  const notifications = () => navigate('NotificationScreen', {});

  const willParticipate = () => {
    SetStoreData('PARTICIPATE', 'true').then(() => {
      LocationServices.start();
      BroadcastingServices.start();
    });

    // Check and see if they actually authorized in the system dialog.
    // If not, stop services and set the state to !isLogging
    // Fixes tripleblindmarket/private-kit#129
    BackgroundGeolocation.checkStatus(({ authorization }) => {
      if (authorization === BackgroundGeolocation.AUTHORIZED) {
        setIsLogging(true);
      } else if (authorization === BackgroundGeolocation.NOT_AUTHORIZED) {
        LocationServices.stop();
        BroadcastingServices.stop();
        setIsLogging(false);
      }
    });
  };

  const setOptOut = () => {
    LocationServices.stop();
    BroadcastingServices.stop();
    setIsLogging(false);
  };

  return (
    <View style={styles.container}>
      <OverlapScreen />
      <TouchableOpacity
        style={styles.searchInput}
        onPress={() => {
          navigate('SearchAddress', {});
        }}>
        <Text style={{ color: '#2E4874' }}>Search location or zip code</Text>
      </TouchableOpacity>
      <SlidingUpPanel
        draggableRange={{
          top: panelHeight,
          bottom: 80,
        }}
        showBackdrop={false}
        containerStyle={styles.panelContainer}
        minimumDistanceThreshold={10}
        friction={50}>
        <View style={styles.bottomDrawer}>
          <View style={styles.ovalWrapper}>
            <View style={styles.oval} />
          </View>
          <View
            style={{ flexDirection: 'row', justifyContent: 'space-between' }}>
            <Text
              style={{
                fontFamily: 'OpenSans-SemiBold',
                fontSize: 17,
                color: '#000',
              }}>
              {isLogging
                ? 'Stop logging my location'
                : 'Start logging my location'}
            </Text>
            <View style={{ paddingRight: 20, height: 40, marginTop: 5 }}>
              <ToggleSwitch
                isOn={isLogging}
                onColor='#2E4874'
                offColor='#2E4874'
                onToggle={isOn => (isOn ? willParticipate() : setOptOut())}
              />
            </View>
          </View>
          <Text
            style={{
              fontFamily: 'OpenSans-Regular',
              fontSize: 13,
              color: '#2E4874',
            }}>
            {isLogging ? 'Your location data is being logged and shared' : ''}
          </Text>
          <View
            style={{ height: 0.3, backgroundColor: 'gray', marginTop: 15 }}
          />
          <View
            style={{
              flexDirection: 'row',
              justifyContent: 'space-around',
              marginTop: 20,
            }}>
            <View style={{ justifyContent: 'center', alignSelf: 'center' }}>
              <View>
                <Text
                  style={{
                    fontFamily: 'OpenSans-Regular',
                    fontSize: 13,
                    color: '#2E4874',
                  }}>
                  BlockList
                </Text>
                <Text
                  style={{
                    fontFamily: 'OpenSans-Regular',
                    fontSize: 13,
                    color: '#2E4874',
                  }}>
                  location
                </Text>
              </View>
            </View>
            <View style={{ justifyContent: 'center', alignSelf: 'center' }}>
              <Text
                style={{
                  fontFamily: 'OpenSans-Regular',
                  fontSize: 13,
                  color: '#2E4874',
                }}>
                Activity
              </Text>
              <Text
                style={{
                  fontFamily: 'OpenSans-Regular',
                  fontSize: 13,
                  color: '#2E4874',
                }}>
                Log
              </Text>
            </View>
          </View>
        </View>
      </SlidingUpPanel>
    </View>
  );
};

const styles = StyleSheet.create({
  // Container covers the entire screen
  container: {
    flex: 1,
    alignItems: 'center',
  },
  searchInput: {
    zIndex: 999,
    position: 'absolute',
    top: 0,
    backgroundColor: '#fff',
    padding: 20,
    width: '95%',
    borderRadius: 14,
    marginTop: 10,
    shadowColor: '#B0C6E2',
    shadowOffset: {
      width: 0,
      height: 12,
    },
    shadowOpacity: 0.58,
    shadowRadius: 16.0,
    elevation: 24,
  },
  ovalWrapper: { alignItems: 'center', width: '100%', paddingBottom: 7 },
  oval: {
    width: 40,
    height: 7,
    backgroundColor: '#CAD2D3',
    borderRadius: 40,
  },
  bottomDrawer: {
    paddingTop: 10,
    paddingLeft: 20,
    paddingRight: 20,
    paddingBottom: 20,
    backgroundColor: '#fff',
    borderRadius: 15,
  },
  panelContainer: {
    zIndex: 9999,
    overflow: 'hidden',

    margin: 15,
  },
});

export default LocationTracking;
