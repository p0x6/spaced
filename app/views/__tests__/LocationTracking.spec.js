import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import LocationTracking from '../MainScreen';

it('renders correctly', async () => {
  const tree = renderer
    .create(<LocationTracking />)
    .toJSON();
  await expect(tree).toMatchSnapshot();
});
