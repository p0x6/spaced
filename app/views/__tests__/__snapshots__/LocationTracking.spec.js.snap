// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFF",
      "color": "#000",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "center",
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "space-between",
        "width": "90%",
      }
    }
  >
    <View>
      <Menu
        style={
          Object {
            "alignSelf": "flex-end",
            "marginTop": "5.5%",
            "position": "absolute",
            "zIndex": 10,
          }
        }
      >
        <MenuTrigger
          style={
            Object {
              "marginTop": 14,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../app/assets/images/kebabIcon.png",
              }
            }
            style={
              Object {
                "height": 28,
                "padding": 14,
                "width": 15,
              }
            }
          />
        </MenuTrigger>
        <MenuOptions>
          <MenuOption
            onSelect={[Function]}
          >
            <Text
              style={
                Object {
                  "fontSize": 14,
                  "fontWeight": "400",
                  "padding": 10,
                }
              }
            >
              Licenses
            </Text>
          </MenuOption>
          <MenuOption
            onSelect={[Function]}
          >
            <Text
              style={
                Object {
                  "fontSize": 14,
                  "fontWeight": "400",
                  "padding": 10,
                }
              }
            >
              Notifications
            </Text>
          </MenuOption>
        </MenuOptions>
      </Menu>
      <Text
        style={
          Object {
            "fontSize": 38,
            "fontWeight": "700",
            "marginTop": "7%",
            "padding": 0,
            "textAlign": "center",
          }
        }
      >
        Private Kit
      </Text>
      <View
        style={
          Object {
            "flex": 6,
            "justifyContent": "flex-start",
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../app/assets/images/PKLogo.png",
            }
          }
          style={
            Object {
              "alignSelf": "center",
              "height": 164.4,
              "marginBottom": 30,
              "marginTop": 15,
              "opacity": 0.3,
              "width": 132,
            }
          }
        />
        <View
          style={
            Object {
              "alignSelf": "center",
              "flexDirection": "row",
              "justifyContent": "center",
              "width": 525,
            }
          }
        >
          <BVLinearGradient
            colors={
              Array [
                4278531739,
                4279786451,
              ]
            }
            endPoint={
              Object {
                "x": 1,
                "y": 0,
              }
            }
            locations={null}
            startPoint={
              Object {
                "x": 0,
                "y": 0,
              }
            }
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderColor": "rgba(255,255,255,0.7)",
                  "borderRadius": 12,
                  "justifyContent": "center",
                  "paddingVertical": 8,
                  "width": "100%",
                },
              ]
            }
          >
            <View
              accessible={true}
              focusable={true}
              isTVSelectable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "borderColor": "rgba(255,255,255,0.7)",
                  "borderRadius": 12,
                  "justifyContent": "center",
                  "opacity": 1,
                  "paddingVertical": 8,
                  "width": "100%",
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#ffffff",
                    "fontSize": 16,
                    "fontWeight": "700",
                    "letterSpacing": 0,
                    "lineHeight": 19,
                    "textAlign": "center",
                  }
                }
              >
                Start tracking
              </Text>
            </View>
          </BVLinearGradient>
        </View>
        <Text
          style={
            Object {
              "color": "#6A6A6A",
              "fontSize": 13,
              "fontWeight": "500",
              "marginBottom": "2%",
              "marginLeft": "12%",
              "marginRight": "12%",
              "marginTop": "2%",
              "textAlign": "center",
            }
          }
        >
          You'll need to allow location permissions
        </Text>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#15D09B",
            "borderRadius": 12,
            "paddingVertical": 8,
            "width": 525,
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "center",
              "marginTop": "2%",
            }
          }
        >
          <SvgXml
            height={15}
            width={15}
            xml="<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<svg width=\\"31px\\" height=\\"38px\\" viewBox=\\"0 0 31 38\\" version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\">
    <!-- Generator: Sketch 64 (93537) - https://sketch.com -->
    <title>twemoji-locked</title>
    <desc>Created with Sketch.</desc>
    <g id=\\"Page-1\\" stroke=\\"none\\" stroke-width=\\"1\\" fill=\\"none\\" fill-rule=\\"evenodd\\">
        <g id=\\"Welcome\\" transform=\\"translate(-58.000000, -447.000000)\\" fill-rule=\\"nonzero\\">
            <g id=\\"twemoji-locked\\" transform=\\"translate(53.000000, 444.000000)\\">
                <g id=\\"Icon\\">
                    <path d=\\"M20.5,3.41666667 C14.2099167,3.41666667 9.11111111,8.51547222 9.11111111,14.8055556 L9.11111111,26.1944444 L13.6666667,26.1944444 L13.6666667,14.8055556 C13.6666667,11.0316098 16.7260542,7.97222222 20.5,7.97222222 C24.2739458,7.97222222 27.3333333,11.0316098 27.3333333,14.8055556 L27.3333333,26.1944444 L31.8888889,26.1944444 L31.8888889,14.8055556 C31.8888889,8.51547222 26.7900833,3.41666667 20.5,3.41666667 Z\\" id=\\"Icon-Path\\" fill=\\"#AAB8C2\\"></path>
                    <path d=\\"M35.3055556,36.4444444 C35.3055556,38.9604083 33.2659639,41 30.75,41 L10.25,41 C7.73403614,41 5.69444444,38.9604083 5.69444444,36.4444444 L5.69444444,22.7777778 C5.69444444,20.2618139 7.73403614,18.2222222 10.25,18.2222222 L30.75,18.2222222 C33.2659639,18.2222222 35.3055556,20.2618139 35.3055556,22.7777778 L35.3055556,36.4444444 Z\\" id=\\"Icon-Path\\" fill=\\"#FFAC33\\"></path>
                </g>
                <rect id=\\"ViewBox\\" x=\\"0\\" y=\\"0\\" width=\\"41\\" height=\\"41\\"></rect>
            </g>
        </g>
    </g>
</svg>"
          />
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 16,
                "fontWeight": "600",
                "marginLeft": 5,
                "textAlign": "center",
              }
            }
          >
            Privacy Info
          </Text>
        </View>
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 12,
              "margin": "2%",
              "marginLeft": "5%",
              "marginRight": "5%",
              "textAlign": "center",
            }
          }
        >
          Your location history stays on your phone and is never shared without your consent
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "fontSize": 12,
            "fontWeight": "600",
            "padding": 4,
            "paddingBottom": 10,
            "textAlign": "center",
          }
        }
      >
        <Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#6A6A6A",
                  "fontSize": 12,
                  "fontWeight": "400",
                  "marginBottom": "2%",
                  "marginLeft": "12%",
                  "marginRight": "12%",
                  "marginTop": "2%",
                  "textAlign": "center",
                },
                Object {
                  "marginLeft": 0,
                  "marginRight": 0,
                },
              ]
            }
          >
            For more info visit 
             
          </Text>
          <Text
            onPress={[Function]}
            style={
              Array [
                Object {
                  "color": "#6A6A6A",
                  "fontSize": 12,
                  "fontWeight": "400",
                  "marginBottom": "2%",
                  "marginLeft": "12%",
                  "marginRight": "12%",
                  "marginTop": "2%",
                  "textAlign": "center",
                },
                Object {
                  "color": "#007AFF",
                  "marginLeft": 0,
                  "marginRight": 0,
                },
              ]
            }
          >
            privatekit.mit.edu
          </Text>
        </Text>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
